client/client-lib/src/api.rs

```diff
@@ -149,33 +159,66 @@ impl WsFederationApi<WsClient> {
     /// Creates a new API client
     pub async fn new(max_evil: usize, members: Vec<(PeerId, String)>) -> Self {
         WsFederationApi {
-            clients: futures::stream::iter(members)
-                .then(|(peer, url)| async move {
-                    // TODO: reconnect to peers on disconnect
-                    (
-                        peer,
-                        WsClientBuilder::default()
-                            .build(url)
-                            .await
-                            .expect("unable to connect to server"),
-                    )
+            members: members
+                .into_iter()
+                .map(|(peer_id, url)| FederationMember {
+                    peer_id,
+                    url,
+                    client: RwLock::new(None),
                 })
-                .collect()
-                .await,
+                .collect(),
             max_evil,
         }
     }
 }
 
-impl<C: ClientT> WsFederationApi<C> {
+impl WsFederationApi<WsClient> {
     pub async fn request<P: serde::Serialize, R: serde::de::DeserializeOwned>(
         &self,
         method: &str,
         param: P,
     ) -> Result<R> {
-        let params = [serde_json::to_value(param).expect("encoding error")];
-        let mut requests = futures::stream::iter(&self.clients).then(|(_id, client)| {
-            client.request::<R>(method, Some(jsonrpsee_types::ParamsSer::ArrayRef(&params)))
+        let params = Arc::new([serde_json::to_value(param).expect("encoding error")]);
+        let mut requests = futures::stream::iter(&self.members).then(|member| {
+            let params = params.clone();
+            Box::pin(async move {
+                let params = Some(jsonrpsee_types::ParamsSer::ArrayRef(&*params));
```

Comment:
This could be moved out of the boxed future so we don't need to clone `params` just to create *another* heap object derived from it and throw the original copy away.

Reply:
inner params borrows from this params clone.

1. serialize params to `JsonValue`
2. create `Arc<JsonValue>`
3. clone Arc for each peer
4. peer request borrows from the Arc

Reply:
Ah, didn't notice the `Arc`! Let's leave it like that, idk what is more efficient: cloning the entire `ParamsSer::Array(vec![…])` struct or using `Arc`+`ParamsSer::ArrayRef` and it's not worth benchmarking rn.

Reply:
didn't need Arc,
it was just due to using stream instead of iterator